components/datetime/src/any/zoned_datetime.rs

```diff
@@ -0,0 +1,333 @@
+/// let zdtf = ZonedDateTimeFormat::<Gregorian>::try_new(
+///     locale!("en"),
+///     &date_provider,
+///     &zone_provider,
+///     &plural_provider,
+///     &decimal_provider,
+///     &options.into(),
+///     &TimeZoneFormatOptions::default(),
+/// )
+/// .expect("Failed to create DateTimeFormat instance.");
+///
+/// let zoned_datetime: MockZonedDateTime = "2021-04-08T16:12:37.000-07:00"
+///     .parse()
+///     .expect("Failed to parse zoned datetime");
+///
+/// let value = zdtf.format_to_string(&zoned_datetime);
+/// ```
+pub struct ZonedAnyDateTimeFormat(raw::ZonedDateTimeFormat, AnyCalendar);
+
+impl ZonedAnyDateTimeFormat {
+    /// Constructor that takes a selected [`Locale`], a reference to a [data provider] for
+    /// dates, a [data provider] for time zones, a [data provider] for calendars, and a list of [`DateTimeFormatOptions`].
+    /// It collects all data necessary to format zoned datetime values into the given locale.
+    ///
+    /// This method is **unstable**, more bounds may be added in the future as calendar support is added. It is
+    /// preferable to use a provider that implements `ResourceProvider<D>` for all `D`, and ensure data is loaded as
+    /// appropriate. The [`Self::try_new_with_buffer_provider()`], [`Self::try_new_with_any_provider()`] constructors
+    /// may also be used if compile stability is desired.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu::calendar::Gregorian;
+    /// use icu::datetime::mock::zoned_datetime::MockZonedDateTime;
+    /// use icu::datetime::{DateTimeFormatOptions, any::ZonedAnyDateTimeFormat};
+    /// use icu::locid::Locale;
+    /// use icu::datetime::TimeZoneFormatOptions;
+    /// use icu_provider::inv::InvariantDataProvider;
+    /// use std::str::FromStr;
+    ///
+    /// let provider = icu_testdata::get_provider();
+    ///
+    /// let options = DateTimeFormatOptions::default();
+    /// let locale = Locale::from_str("en-u-ca-gregory").unwrap();
+    ///
+    /// let zdtf = ZonedAnyDateTimeFormat::try_new_unstable(
+    ///     locale,
+    ///     &provider,
+    ///     &provider,
+    ///     &provider,
+    ///     &provider,
+    ///     &provider,
+    ///     &options,
+    ///     &TimeZoneFormatOptions::default(),
+    /// );
+    ///
+    /// assert_eq!(zdtf.is_ok(), true);
+    /// ```
+    ///
+    /// [data provider]: icu_provider
+    #[inline]
+    #[allow(clippy::too_many_arguments)]
+    pub fn try_new_unstable<L, DP, ZP, PP, DEP, CEP>(
+        locale: L,
+        date_provider: &DP,
+        zone_provider: &ZP,
+        plural_provider: &PP,
+        decimal_provider: &DEP,
+        calendar_provider: &CEP,
+        date_time_format_options: &DateTimeFormatOptions,
+        time_zone_format_options: &TimeZoneFormatOptions,
+    ) -> Result<Self, DateTimeFormatError>
+    where
+        L: Into<Locale>,
+        DP: ResourceProvider<DateSymbolsV1Marker>
```

Comment:
Thought: With `try_new_unstable` we should eventually be able to collapse the providers into one and use a forking provider to delegate between them at the call site. @robertbastian tried getting that to work in ForkByKey but it's a bit tricky. It may require a macro to generate the match statement. Not sure if we want to make this change now or later.

Reply:
Yeah. I didn't want to change the API too much in this change.